### ISSUE TITLE
M8 — MNIST CNN to >98%, end to end

### ISSUE THREAD

**Issue by harshitagrawal2O** [harshitagrawal2O/tensorkit#8]
**Milestone 8 of 8** — see [`SPEC.md`](../blob/main/SPEC.md) section 4 for the full definition of done.

**CS subject:** Everything above, integrated · Experimental methodology

**Nobody is working on this.** Comment to claim it. One person per milestone at a time; if a
claim goes quiet for two weeks it is fair game again.

## Scope

Train a CNN on MNIST using only this library. Produce the accuracy curve and a full gradcheck sweep.

## Definition of done

- `examples/train_mnist.py` reaches **>98% test accuracy**, reproducible from a fixed seed
- `BENCHMARKS.md` carries throughput vs PyTorch, peak memory, and a hardware header
- `make purity` still passes — the PyTorch comparison imports torch in `benchmarks/` only

## Where it gets interesting

This is the milestone that proves the previous seven. Also the one where you find out what a pure-NumPy engine actually costs, and have to account for the gap honestly.

## Ground rules

- Do **not** weaken a test to make it pass. If you believe a test is wrong, open a
  [spec-defect issue](../issues/new?template=spec-defect.yml) instead — several tests have
  already been corrected that way.
- Keep the existing docstrings; replace only the `raise NotImplementedError(...)` line.
- `ruff check`, `ruff format --check` and `mypy` must be clean.

Chapter-length background is in [the book](../blob/main/docs/tensorkit-book.pdf).

```bash
python -m pytest -m m8 -q
```
